core/domain/exp_jobs_one_off.py

```diff
@@ -490,3 +493,108 @@ def reduce(key, values):
+        try:
+            exploration.validate()
+        except Exception as e:
+            logging.error(
+                'Exploration %s failed non-strict validation: %s' %
+                (item.id, e))
+            return
+        if (item.states_schema_version <
+                AUDIO_DURATION_SECS_MIN_STATE_SCHEMA_VERSION):
+            yield ('State schema version is not the '
+                   'minimum version expected', item.id)
+            return
+
+        # Go through each exploration state to find voiceover recordings.
+        # For batching them as commits per explorations.
+        commit_cmds = []
+        states_changed_count = 0
+        unchanged_count = 0
+        failed_count = 0
+        for state, state_value in item.states.items():
+            voiceovers_mapping = (state_value['recorded_voiceovers']
+                                  ['voiceovers_mapping'])
+            language_codes_to_audio_metadata = voiceovers_mapping.values()
+            for language_codes in language_codes_to_audio_metadata:
+                for audio_metadata in language_codes.values():
+                    # Get files using the filename.
+                    filename = audio_metadata['filename']
+                    if audio_metadata['duration_secs'] == 0.0:
+                        try:
+                            fs = (fs_domain.AbstractFileSystem(
+                                fs_domain.GcsFileSystem(
+                                    AUDIO_ENTITY_TYPE,
+                                    item.id)))
+                            raw = (
+                                fs.get('%s/%s' % (AUDIO_FILE_PREFIX,
+                                                  filename)))
+                            # Get the audio-duration from file use Mutagen.
+                            tempbuffer = python_utils.string_io()
+                            tempbuffer.write(raw)
+                            tempbuffer.seek(0)
+                            # Loads audio metadata with Mutagen.
+                            audio = mp3.MP3(tempbuffer)
+                            tempbuffer.close()
+                            # Fetch the audio file duration from the Mutagen
+                            # metadata.
+                            audio_metadata['duration_secs'] = (
+                                audio.info.length)
+                            states_changed_count += 1
+                        except Exception as e:
+                            logging.error(
+                                'MP3 audio file exception for file %s ,'
+                                'EXP_ID: %s,'
+                                'STATE_NAME: %s,'
+                                'REASON: %s' %
+                                (filename, item.id, state, e))
+                            failed_count += 1
+                    else:
+                        unchanged_count += 1
+            if states_changed_count > 0:
```

Comment:
states_changed_count is outside the for loop, but this is inside the for loop. The logic here seems wrong?

E.g. if the first state is changed and the second isn't, this code will still process the second state. 

(Please also add a backend test that would catch this sort of thing.)

Reply:
Yeah you're right.

Reply:
Ok so for every change of state do we make that a commit? Or do we batch all of the state changes for as a single commit for the exploration?

Reply:
Fixed the logic such that the commits are made only when the state has at least one change.

Reply:
Done.